{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Genomic Data for Variant Pathogenicity\n",
    "This notebook reads the vcf file containing ClinVar data and outputs a vcf file that contains the right information to run ANNOVAR and, eventually reach the table templated format provide in FH-EARLY for the genomic data.\n",
    "\n",
    "### To download the ClinVar data:\n",
    "Go to https://ftp.ncbi.nlm.nih.gov/pub/clinvar/vcf_GRCh38/ (Our version is clinvar_20260208.vcf.gz)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import packages\n",
    "import pandas as pd\n",
    "from cyvcf2 import VCF # https://github.com/brentp/cyvcf2/tree/main"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>id</th>\n",
       "      <th>qual</th>\n",
       "      <th>clnsig</th>\n",
       "      <th>geneinfo</th>\n",
       "      <th>gt_types</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>19</td>\n",
       "      <td>11087729</td>\n",
       "      <td>ACCACGCCCGGCTAATTTTTTGTATTTTTTTTTAGTAGAGGTGGGG...</td>\n",
       "      <td>A</td>\n",
       "      <td>694275</td>\n",
       "      <td>None</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>LDLR:3949|LDLR-AS1:115271120</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>19</td>\n",
       "      <td>11089263</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>430740</td>\n",
       "      <td>None</td>\n",
       "      <td>Uncertain_significance</td>\n",
       "      <td>LDLR:3949|LDLR-AS1:115271120</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>19</td>\n",
       "      <td>11089281</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>250925</td>\n",
       "      <td>None</td>\n",
       "      <td>Benign/Likely_benign</td>\n",
       "      <td>LDLR:3949|LDLR-AS1:115271120</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>19</td>\n",
       "      <td>11089283</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>4069787</td>\n",
       "      <td>None</td>\n",
       "      <td>Uncertain_significance</td>\n",
       "      <td>LDLR:3949|LDLR-AS1:115271120</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>19</td>\n",
       "      <td>11089309</td>\n",
       "      <td>CAG</td>\n",
       "      <td>C</td>\n",
       "      <td>3628882</td>\n",
       "      <td>None</td>\n",
       "      <td>Uncertain_significance</td>\n",
       "      <td>LDLR:3949|LDLR-AS1:115271120</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4319</th>\n",
       "      <td>19</td>\n",
       "      <td>11133635</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>328126</td>\n",
       "      <td>None</td>\n",
       "      <td>Conflicting_classifications_of_pathogenicity</td>\n",
       "      <td>LDLR:3949</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4320</th>\n",
       "      <td>19</td>\n",
       "      <td>11133666</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>890682</td>\n",
       "      <td>None</td>\n",
       "      <td>Uncertain_significance</td>\n",
       "      <td>LDLR:3949</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4321</th>\n",
       "      <td>19</td>\n",
       "      <td>11133681</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>328127</td>\n",
       "      <td>None</td>\n",
       "      <td>Uncertain_significance</td>\n",
       "      <td>LDLR:3949</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4322</th>\n",
       "      <td>19</td>\n",
       "      <td>11133682</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>328128</td>\n",
       "      <td>None</td>\n",
       "      <td>Uncertain_significance</td>\n",
       "      <td>LDLR:3949</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4323</th>\n",
       "      <td>19</td>\n",
       "      <td>11133700</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>328129</td>\n",
       "      <td>None</td>\n",
       "      <td>Uncertain_significance</td>\n",
       "      <td>LDLR:3949</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4324 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     chrom       pos                                                ref alt  \\\n",
       "0       19  11087729  ACCACGCCCGGCTAATTTTTTGTATTTTTTTTTAGTAGAGGTGGGG...   A   \n",
       "1       19  11089263                                                  C   G   \n",
       "2       19  11089281                                                  G   T   \n",
       "3       19  11089283                                                  C   T   \n",
       "4       19  11089309                                                CAG   C   \n",
       "...    ...       ...                                                ...  ..   \n",
       "4319    19  11133635                                                  C   G   \n",
       "4320    19  11133666                                                  C   G   \n",
       "4321    19  11133681                                                  C   T   \n",
       "4322    19  11133682                                                  G   T   \n",
       "4323    19  11133700                                                  G   T   \n",
       "\n",
       "           id  qual                                        clnsig  \\\n",
       "0      694275  None                                    Pathogenic   \n",
       "1      430740  None                        Uncertain_significance   \n",
       "2      250925  None                          Benign/Likely_benign   \n",
       "3     4069787  None                        Uncertain_significance   \n",
       "4     3628882  None                        Uncertain_significance   \n",
       "...       ...   ...                                           ...   \n",
       "4319   328126  None  Conflicting_classifications_of_pathogenicity   \n",
       "4320   890682  None                        Uncertain_significance   \n",
       "4321   328127  None                        Uncertain_significance   \n",
       "4322   328128  None                        Uncertain_significance   \n",
       "4323   328129  None                        Uncertain_significance   \n",
       "\n",
       "                          geneinfo gt_types  \n",
       "0     LDLR:3949|LDLR-AS1:115271120       []  \n",
       "1     LDLR:3949|LDLR-AS1:115271120       []  \n",
       "2     LDLR:3949|LDLR-AS1:115271120       []  \n",
       "3     LDLR:3949|LDLR-AS1:115271120       []  \n",
       "4     LDLR:3949|LDLR-AS1:115271120       []  \n",
       "...                            ...      ...  \n",
       "4319                     LDLR:3949       []  \n",
       "4320                     LDLR:3949       []  \n",
       "4321                     LDLR:3949       []  \n",
       "4322                     LDLR:3949       []  \n",
       "4323                     LDLR:3949       []  \n",
       "\n",
       "[4324 rows x 9 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vcf_og = VCF('clinvar_20260208.vcf.gz')\n",
    "rows = []\n",
    "for var in vcf_og('19:11087732-11133700'):\n",
    "    # TODO: select necessary information\n",
    "    rows.append({\n",
    "        'chrom': var.CHROM,\n",
    "        'pos': var.POS,  # 1-based\n",
    "        'ref': var.REF,\n",
    "        'alt': ','.join(var.ALT),\n",
    "        'id': var.ID,\n",
    "        'qual': var.QUAL,\n",
    "        'clnsig': var.INFO.get('CLNSIG', ''),\n",
    "        'geneinfo': var.INFO.get('GENEINFO', ''),\n",
    "        'gt_types': var.gt_types.tolist()\n",
    "    })\n",
    "\n",
    "df = pd.DataFrame(rows)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "geneinfo\n",
       "LDLR:3949                           4062\n",
       "LDLR:3949|LDLR-AS1:115271120         220\n",
       "LDLR:3949|MIR6886:102465534           41\n",
       "LDLR:3949|LOC126862855:126862855       1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['geneinfo'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* only samples encoding LDLR variants\n",
    "* id is unique\n",
    "* pos is almost unique, there are some variants in the same position"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>id</th>\n",
       "      <th>qual</th>\n",
       "      <th>clnsig</th>\n",
       "      <th>geneinfo</th>\n",
       "      <th>gt_types</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1201</th>\n",
       "      <td>19</td>\n",
       "      <td>11105575</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>2094541</td>\n",
       "      <td>None</td>\n",
       "      <td>Likely_benign</td>\n",
       "      <td>LDLR:3949</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1202</th>\n",
       "      <td>19</td>\n",
       "      <td>11105575</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>251370</td>\n",
       "      <td>None</td>\n",
       "      <td>Likely_pathogenic</td>\n",
       "      <td>LDLR:3949</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1203</th>\n",
       "      <td>19</td>\n",
       "      <td>11105575</td>\n",
       "      <td>G</td>\n",
       "      <td>GGACAAA</td>\n",
       "      <td>977984</td>\n",
       "      <td>None</td>\n",
       "      <td>Likely_pathogenic</td>\n",
       "      <td>LDLR:3949</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1204</th>\n",
       "      <td>19</td>\n",
       "      <td>11105575</td>\n",
       "      <td>G</td>\n",
       "      <td>GGACAAATCT</td>\n",
       "      <td>251377</td>\n",
       "      <td>None</td>\n",
       "      <td>Likely_pathogenic</td>\n",
       "      <td>LDLR:3949</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1205</th>\n",
       "      <td>19</td>\n",
       "      <td>11105575</td>\n",
       "      <td>G</td>\n",
       "      <td>GGACAAATCTGAC</td>\n",
       "      <td>251376</td>\n",
       "      <td>None</td>\n",
       "      <td>Likely_pathogenic</td>\n",
       "      <td>LDLR:3949</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1206</th>\n",
       "      <td>19</td>\n",
       "      <td>11105575</td>\n",
       "      <td>GGAC</td>\n",
       "      <td>G</td>\n",
       "      <td>2125438</td>\n",
       "      <td>None</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>LDLR:3949</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1207</th>\n",
       "      <td>19</td>\n",
       "      <td>11105575</td>\n",
       "      <td>GGACAAATCTGACGA</td>\n",
       "      <td>AACTGCGGTAAACTGCGGTAAACT</td>\n",
       "      <td>430761</td>\n",
       "      <td>None</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>LDLR:3949</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     chrom       pos              ref                       alt       id  \\\n",
       "1201    19  11105575                G                         A  2094541   \n",
       "1202    19  11105575                G                         C   251370   \n",
       "1203    19  11105575                G                   GGACAAA   977984   \n",
       "1204    19  11105575                G                GGACAAATCT   251377   \n",
       "1205    19  11105575                G             GGACAAATCTGAC   251376   \n",
       "1206    19  11105575             GGAC                         G  2125438   \n",
       "1207    19  11105575  GGACAAATCTGACGA  AACTGCGGTAAACTGCGGTAAACT   430761   \n",
       "\n",
       "      qual             clnsig   geneinfo gt_types  \n",
       "1201  None      Likely_benign  LDLR:3949       []  \n",
       "1202  None  Likely_pathogenic  LDLR:3949       []  \n",
       "1203  None  Likely_pathogenic  LDLR:3949       []  \n",
       "1204  None  Likely_pathogenic  LDLR:3949       []  \n",
       "1205  None  Likely_pathogenic  LDLR:3949       []  \n",
       "1206  None         Pathogenic  LDLR:3949       []  \n",
       "1207  None         Pathogenic  LDLR:3949       []  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[df['pos']==11105575]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>id</th>\n",
       "      <th>clnsig</th>\n",
       "      <th>geneinfo</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>19</td>\n",
       "      <td>11087729</td>\n",
       "      <td>ACCACGCCCGGCTAATTTTTTGTATTTTTTTTTAGTAGAGGTGGGG...</td>\n",
       "      <td>A</td>\n",
       "      <td>694275</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>LDLR:3949|LDLR-AS1:115271120</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>19</td>\n",
       "      <td>11089263</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>430740</td>\n",
       "      <td>Uncertain_significance</td>\n",
       "      <td>LDLR:3949|LDLR-AS1:115271120</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>19</td>\n",
       "      <td>11089281</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>250925</td>\n",
       "      <td>Benign/Likely_benign</td>\n",
       "      <td>LDLR:3949|LDLR-AS1:115271120</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>19</td>\n",
       "      <td>11089283</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>4069787</td>\n",
       "      <td>Uncertain_significance</td>\n",
       "      <td>LDLR:3949|LDLR-AS1:115271120</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>19</td>\n",
       "      <td>11089309</td>\n",
       "      <td>CAG</td>\n",
       "      <td>C</td>\n",
       "      <td>3628882</td>\n",
       "      <td>Uncertain_significance</td>\n",
       "      <td>LDLR:3949|LDLR-AS1:115271120</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom       pos                                                ref alt  \\\n",
       "0    19  11087729  ACCACGCCCGGCTAATTTTTTGTATTTTTTTTTAGTAGAGGTGGGG...   A   \n",
       "1    19  11089263                                                  C   G   \n",
       "2    19  11089281                                                  G   T   \n",
       "3    19  11089283                                                  C   T   \n",
       "4    19  11089309                                                CAG   C   \n",
       "\n",
       "        id                  clnsig                      geneinfo  \n",
       "0   694275              Pathogenic  LDLR:3949|LDLR-AS1:115271120  \n",
       "1   430740  Uncertain_significance  LDLR:3949|LDLR-AS1:115271120  \n",
       "2   250925    Benign/Likely_benign  LDLR:3949|LDLR-AS1:115271120  \n",
       "3  4069787  Uncertain_significance  LDLR:3949|LDLR-AS1:115271120  \n",
       "4  3628882  Uncertain_significance  LDLR:3949|LDLR-AS1:115271120  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "to_remove = ['gt_types', 'qual']\n",
    "df_v2 = df.drop(columns=to_remove)\n",
    "df_v2.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# information needed for ANNOVAR\n",
    "1. chromosome\n",
    "2. start position (pos)\n",
    "3. end position - TODO\n",
    "4. ref nucleotide (ref)\n",
    "5. observed nucleotide (alt)\n",
    "\n",
    "## TO CHECK: I think ANNOVAR is only for missense (1 nucleotide)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "catt",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.14.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
