{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "from cyvcf2 import VCF\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['ALT', 'CHROM', 'FILTER', 'FILTERS', 'FORMAT', 'ID', 'INFO', 'POS', 'QUAL', 'REF', '__class__', '__delattr__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__getstate__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__le__', '__lt__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__setstate__', '__sizeof__', '__str__', '__subclasshook__', 'aaf', 'call_rate', 'end', 'format', 'genotype', 'genotypes', 'gt_alt_depths', 'gt_alt_freqs', 'gt_bases', 'gt_depths', 'gt_phases', 'gt_phred_ll_het', 'gt_phred_ll_homalt', 'gt_phred_ll_homref', 'gt_quals', 'gt_ref_depths', 'gt_types', 'is_deletion', 'is_indel', 'is_mnp', 'is_snp', 'is_sv', 'is_transition', 'nucl_diversity', 'num_called', 'num_het', 'num_hom_alt', 'num_hom_ref', 'num_unknown', 'ploidy', 'relatedness', 'set_format', 'set_pos', 'start', 'var_subtype', 'var_type']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[W::vcf_parse] Contig '1' is not defined in the header. (Quick workaround: index the file with tabix.)\n"
     ]
    }
   ],
   "source": [
    "vcf = VCF(\"clinvar_20260201.vcf\")\n",
    "variant = next(iter(vcf))  # get first variant\n",
    "\n",
    "print(dir(variant))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[W::vcf_parse] Contig '2' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '3' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '4' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '5' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '6' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '7' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '8' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '9' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '10' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '11' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '12' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '13' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '14' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '15' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '16' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '17' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '18' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '19' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '20' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '21' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '22' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'X' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'Y' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'MT' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'NT_113889.1' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'NT_113891.3' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'NT_187576.1' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'NT_187633.1' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'NT_187661.1' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'NT_187693.1' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'NW_009646201.1' is not defined in the header. (Quick workaround: index the file with tabix.)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>id</th>\n",
       "      <th>qual</th>\n",
       "      <th>filter</th>\n",
       "      <th>info</th>\n",
       "      <th>gt_types</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>69134</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>2205837</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>{'ALLELEID': 2193183, 'CLNDISDB': 'MedGen:CN16...</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>69241</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>4562067</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>{'ALLELEID': 4679177, 'CLNDISDB': 'MedGen:CN16...</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>69308</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>3925305</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>{'ALLELEID': 4039319, 'CLNDISDB': 'MedGen:CN16...</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>69314</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>3205580</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>{'ALLELEID': 3374047, 'CLNDISDB': 'MedGen:CN16...</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>69404</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>3925306</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>{'ALLELEID': 4039320, 'CLNDISDB': 'MedGen:CN16...</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom    pos ref alt       id  qual filter  \\\n",
       "0     1  69134   A   G  2205837  None   None   \n",
       "1     1  69241   C   T  4562067  None   None   \n",
       "2     1  69308   A   G  3925305  None   None   \n",
       "3     1  69314   T   G  3205580  None   None   \n",
       "4     1  69404   T   C  3925306  None   None   \n",
       "\n",
       "                                                info gt_types  \n",
       "0  {'ALLELEID': 2193183, 'CLNDISDB': 'MedGen:CN16...       []  \n",
       "1  {'ALLELEID': 4679177, 'CLNDISDB': 'MedGen:CN16...       []  \n",
       "2  {'ALLELEID': 4039319, 'CLNDISDB': 'MedGen:CN16...       []  \n",
       "3  {'ALLELEID': 3374047, 'CLNDISDB': 'MedGen:CN16...       []  \n",
       "4  {'ALLELEID': 4039320, 'CLNDISDB': 'MedGen:CN16...       []  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rows = []\n",
    "for var in vcf:\n",
    "    rows.append({\n",
    "        \"chrom\": var.CHROM,\n",
    "        \"pos\": var.POS,              # 1-based\n",
    "        \"ref\": var.REF,\n",
    "        \"alt\": \",\".join(var.ALT),\n",
    "        \"id\": var.ID,\n",
    "        \"qual\": var.QUAL,\n",
    "        \"filter\": var.FILTER,\n",
    "        \"info\": dict(var.INFO),      # INFO fields as dict\n",
    "        # genotype example (per-sample GT types)\n",
    "        # 0=homozygous ref, 1=het, 2=homozygous alt, 3=unknown\n",
    "        \"gt_types\": var.gt_types.tolist(),\n",
    "    })\n",
    "\n",
    "df = pd.DataFrame(rows)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(4281769, 9)"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'AF_ESP': 7.999999797903001e-05,\n",
       " 'ALLELEID': 1980399,\n",
       " 'CLNDISDB': 'MedGen:C3661900',\n",
       " 'CLNDN': 'not_provided',\n",
       " 'CLNHGVS': 'NC_000001.11:g.944006C>T',\n",
       " 'CLNREVSTAT': 'criteria_provided,_single_submitter',\n",
       " 'CLNSIG': 'Likely_benign',\n",
       " 'CLNSIGSCV': 'SCV002967043',\n",
       " 'CLNVC': 'single_nucleotide_variant',\n",
       " 'CLNVCSO': 'SO:0001483',\n",
       " 'CLNVI': 'ClinGen:CA16728409',\n",
       " 'GENEINFO': 'SAMD11:148398',\n",
       " 'MC': 'SO:0001819|synonymous_variant',\n",
       " 'ORIGIN': '1',\n",
       " 'RS': '376249278'}"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['info'][1000]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>id</th>\n",
       "      <th>qual</th>\n",
       "      <th>filter</th>\n",
       "      <th>info</th>\n",
       "      <th>gt_types</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>65595</th>\n",
       "      <td>1</td>\n",
       "      <td>21812458</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>2353725</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>{'ALLELEID': 2340390, 'CLNDISDB': 'MedGen:CN16...</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65596</th>\n",
       "      <td>1</td>\n",
       "      <td>21812482</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>3537752</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>{'ALLELEID': 3698972, 'CLNDISDB': 'MedGen:CN16...</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65597</th>\n",
       "      <td>1</td>\n",
       "      <td>21812487</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>4046759</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>{'ALLELEID': 4163335, 'CLNDISDB': 'MedGen:CN16...</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65598</th>\n",
       "      <td>1</td>\n",
       "      <td>21812522</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>3866815</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>{'ALLELEID': 3984332, 'CLNDISDB': 'MedGen:CN16...</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65599</th>\n",
       "      <td>1</td>\n",
       "      <td>21814404</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>2482955</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>{'AF_EXAC': 9.999999747378752e-06, 'ALLELEID':...</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      chrom       pos ref alt       id  qual filter  \\\n",
       "65595     1  21812458   G   C  2353725  None   None   \n",
       "65596     1  21812482   C   A  3537752  None   None   \n",
       "65597     1  21812487   G   T  4046759  None   None   \n",
       "65598     1  21812522   C   A  3866815  None   None   \n",
       "65599     1  21814404   T   A  2482955  None   None   \n",
       "\n",
       "                                                    info gt_types  \n",
       "65595  {'ALLELEID': 2340390, 'CLNDISDB': 'MedGen:CN16...       []  \n",
       "65596  {'ALLELEID': 3698972, 'CLNDISDB': 'MedGen:CN16...       []  \n",
       "65597  {'ALLELEID': 4163335, 'CLNDISDB': 'MedGen:CN16...       []  \n",
       "65598  {'ALLELEID': 3984332, 'CLNDISDB': 'MedGen:CN16...       []  \n",
       "65599  {'AF_EXAC': 9.999999747378752e-06, 'ALLELEID':...       []  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.DataFrame(rows)\n",
    "df = df[df['info'].apply(lambda x: 'LDLR' in str(x.get('GENEINFO', '')) if x else False)]\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "catt",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.14.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
